cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -27,19 +32,139 @@
  */
 public class BasicWorkflowToken implements WorkflowToken {
   private Map<String, Map<String, Long>> mapReduceCounters;
```

Comment:
Since the `tokenValueMap` maintains flatten counter names as `mr.counters.groupname.countername`, we won't be able to reconstruct the group name and counter name from it. I thought of maintaining the map of all counter group names and counter names and then use them to recreate the `mapReduceCounters`, however it will not work if the user has added the custom counter.